instance/models/mixins/storage.py

```diff
@@ -481,26 +479,86 @@ def deprovision_s3(self):
-                    'There was an error trying to remove S3 bucket "%s".',
-                    self.s3_bucket_name
-                )
-        else:
-            self.s3_bucket_name = ""
-            self.save()
+            if e.response.get('Error', {}).get('Code') == 'NoSuchBucket':
+                self.logger.info('Could not delete S3 bucket; bucket not found')
+            else:
+                raise
+
+    def _delete_s3_access_key(self):
+        """
+        Deletes the S3 access key
+        """
+        if not self.s3_access_key:
+            return
 
         try:
-            # Access keys and policies need to be deleted before removing the user
             self.iam.delete_access_key(UserName=self.iam_username, AccessKeyId=self.s3_access_key)
+        except ClientError as e:
+            if e.response.get('Error', {}).get('Code') == 'NoSuchEntity':
+                self.logger.info('Could not delete S3 access key; access key not found')
+            else:
+                raise
+
+    def _delete_s3_user_policy(self):
+        """
+        Deletes the S3 IAM user policy
+        """
+        try:
             self.iam.delete_user_policy(UserName=self.iam_username, PolicyName=USER_POLICY_NAME)
+        except ClientError as e:
+            if e.response.get('Error', {}).get('Code') == 'NoSuchEntity':
+                self.logger.info('Could not delete S3 user policy; user policy not found')
+            else:
+                raise
+
+    def _delete_s3_user(self):
+        """
+        Deletes the S3 IAM user
+        """
+        try:
             self.iam.delete_user(UserName=self.iam_username)
-            self.s3_access_key = ""
-            self.s3_secret_access_key = ""
+        except ClientError as e:
+            if e.response.get('Error', {}).get('Code') == 'NoSuchEntity':
+                self.logger.info('Could not delete S3 user; user not found')
+            else:
+                raise
+
+    def deprovision_s3(self):
+        """
+        Deprovision S3 by deleting S3 bucket and IAM user
+        """
+        self.logger.info('Deprovisioning S3 started.')
+        if (not self.storage_type == self.S3_STORAGE or
+                not (self.s3_access_key or self.s3_secret_access_key or self.s3_bucket_name)):
+            self.logger.info('Deprovisioning S3 finished (nothing to do).')
+            return
+
+        try:
+            self._delete_s3_bucket()
+        except Exception:
+            self.logger.exception(
+                'There was an error trying to remove S3 bucket "%s".',
+                self.s3_bucket_name
+            )
+            raise
+        else:
+            self.s3_bucket_name = ""
             self.save()
-        except ClientError:
+
+        try:
+            # Access keys and policies need to be deleted before removing the user
+            self._delete_s3_access_key()
+            self._delete_s3_user_policy()
+            self._delete_s3_user()
+        except Exception:
```

Comment:
@mtyaka We may need some debug here like which step failed and the error message, right?

Reply:
It should be possible to see which step failed from the stacktrace (we re-raise the exception), and since errors like this would rarely occur, I don't think we need to do anything special.